src/goto-instrument/replace_function_bodies.cpp

```diff
@@ -0,0 +1,367 @@
+          side_effect_expr_nondett(parameter.type().subtype()));
+      }
+    }
+
+    for(auto const &global_id : globals_to_havoc)
+    {
+      auto assign_instruction = function.body.add_instruction();
+      assign_instruction->make_assignment();
+      auto global_sym = ns.lookup(global_id);
+      assign_instruction->code = code_assignt(
+        symbol_exprt(global_sym.name),
+        side_effect_expr_nondett(global_sym.type));
+    }
+    if(function.type.return_type() != void_typet())
+    {
+      auto return_instruction = function.body.add_instruction();
+      return_instruction->make_return();
+      return_instruction->code =
+        code_returnt(side_effect_expr_nondett(function.type.return_type()));
+    }
+    auto end_function_instruction = function.body.add_instruction();
+    end_function_instruction->make_end_function();
+  }
+
+private:
+  const std::vector<irep_idt> globals_to_havoc;
+  std::regex parameters_to_havoc;
+};
+
+static std::vector<std::string>
+tokenize(const std::string &base, char separator)
+{
+  std::vector<std::string> tokens;
+  tokens.push_back("");
+  for(char c : base)
+  {
+    if(c == separator)
+    {
+      tokens.push_back("");
+    }
+    else
+    {
+      tokens.back().push_back(c);
+    }
+  }
+  return tokens;
+}
+
+static void write_to_stream(std::ostream &out)
+{
+}
+
+template <typename T, typename... Ts>
+void write_to_stream(std::ostream &out, const T &arg, const Ts &... args)
+{
+  out << arg;
+  write_to_stream(out, args...);
+}
+
+template <typename... Ts>
+std::string replace_function_body_parse_error(const Ts &... args)
+{
+  std::ostringstream out;
+  write_to_stream(
+    out, "parse error for --replace-function-body-options: ", args...);
+  return out.str();
+}
+
+std::unique_ptr<replace_function_bodiest> replace_function_body_options_factory(
+  const std::string &options,
+  const symbol_tablet &symbol_table)
+{
+  if(options.empty() || options == "nondet-return")
+  {
+    return util_make_unique<havoc_replace_function_bodiest>(
+      std::vector<irep_idt>{}, std::regex{});
+  }
+
+  if(options == "assume-false")
+  {
+    return util_make_unique<assume_false_replace_function_bodiest>();
+  }
+
+  if(options == "assert-false")
+  {
+    return util_make_unique<assert_false_replace_function_bodiest>();
+  }
+
+  if(options == "assert-false-assume-false")
+  {
+    return util_make_unique<
+      assert_false_then_assume_false_replace_function_bodiest>();
+  }
+
+  if(options == "empty")
+  {
+    return util_make_unique<empty_replace_function_bodiest>();
+  }
+
+  const std::vector<std::string> tokens = tokenize(options, ',');
+  if(!tokens.empty() && tokens[0] == "havoc")
+  {
+    std::regex globals_regex;
+    std::regex params_regex;
+    for(std::size_t i = 1; i < tokens.size(); ++i)
+    {
+      const std::vector<std::string> key_value_pair = tokenize(tokens[i], ':');
+      if(key_value_pair.size() != 2)
+      {
+        throw replace_function_body_parse_error(
```

Comment:
Avoid introducing `throw std::string` type statements, consider having a `replace_function_body_exceptiont` for this

Reply:
Should it be a child of std::runtime_error? If yes, should the message include __FILE__ and __LINE__?

Reply:
They both sound like sensible suggestions to me.